File: problems-extend/1401-1500/1429_first-unique-number/solution.ts
class FirstUnique {
    private counts = new Map<number, number>();
    private queue: number[] = [];
    private head = 0;

    constructor(nums: number[]) {
        for (const value of nums) {
            this.add(value);
        }
    }

    showFirstUnique(): number {
        while (this.head < this.queue.length && this.counts.get(this.queue[this.head])! > 1) {
            this.head++;
        }
        return this.head < this.queue.length ? this.queue[this.head] : -1;
    }

    add(value: number): void {
        this.counts.set(value, (this.counts.get(value) || 0) + 1);
        this.queue.push(value);
    }
}

File: problems-extend/1401-1500/1429_first-unique-number/solution.cpp
#include <queue>
#include <unordered_map>
#include <vector>

class FirstUnique {
  public:
    FirstUnique(std::vector<long long> nums) {
        for (long long value : nums) {
            add(value);
        }
    }

    long long showFirstUnique() {
        while (!pending.empty() && counts[pending.front()] > 1) {
            pending.pop();
        }
        return pending.empty() ? -1 : pending.front();
    }

    void add(long long value) {
        counts[value] += 1;
        pending.push(value);
    }

  private:
    std::unordered_map<long long, int> counts;
    std::queue<long long> pending;
};
